libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/manager/AdvancedSyncUpTask.kt

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2019-present, salesforce.com, inc.
+ * All rights reserved.
+ * Redistribution and use of this software in source and binary forms, with or
+ * without modification, are permitted provided that the following conditions
+ * are met:
+ * - Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * - Neither the name of salesforce.com, inc. nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission of salesforce.com, inc.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.mobilesync.manager
+
+import com.salesforce.androidsdk.mobilesync.manager.SyncManager.SyncUpdateCallback
+import com.salesforce.androidsdk.mobilesync.target.AdvancedSyncUpTarget
+import com.salesforce.androidsdk.mobilesync.target.SyncUpTarget
+import com.salesforce.androidsdk.mobilesync.util.SyncOptions
+import com.salesforce.androidsdk.mobilesync.util.SyncState
+import com.salesforce.androidsdk.mobilesync.util.SyncState.MergeMode
+import com.salesforce.androidsdk.smartstore.store.SmartStore
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.IOException
+
+/**
+ * Runnable class responsible for running a sync up that uses and AdvancedSyncUpTarget
+ */
+class AdvancedSyncUpTask(syncManager: SyncManager, sync: SyncState, callback: SyncUpdateCallback?) :
+    SyncUpTask(syncManager, sync, callback) {
+    @Throws(JSONException::class, IOException::class)
+    override fun syncUp(
+        sync: SyncState,
+        callback: SyncUpdateCallback?,
+        dirtyRecordIds: List<String>
+    ) {
+        val soupName = sync.soupName
+        val target = sync.target as SyncUpTarget
+        val options = sync.options
+        val totalSize = dirtyRecordIds.size
+        val maxBatchSize = (target as AdvancedSyncUpTarget).maxBatchSize
+        val batch: MutableList<JSONObject> = ArrayList()
```

Comment:
Optional: A lot of specific collection types can switch to Kotlin's collections, such as  `mutableListOf<JSONObject>()` in this case.